src/net.cpp

```diff
@@ -2009,25 +2019,46 @@ bool CConnman::InactivityCheck(const CNode& node) const
     return false;
 }
 
+bool CConnman::ShouldTryToSend(NodeId node_id) const
+{
+    AssertLockNotHeld(m_nodes_mutex);
+
+    CNode* node{GetNodeById(node_id)};
+    if (node == nullptr) {
+        return false;
+    }
+    LOCK(node->cs_vSend);
+    // Sending is possible if either there are bytes to send right now, or if there will be
+    // once a potential message from vSendMsg is handed to the transport. GetBytesToSend
+    // determines both of these in a single call.
+    const auto& [to_send, more, _msg_type] = node->m_transport->GetBytesToSend(!node->vSendMsg.empty());
+    return !to_send.empty() || more;
+}
+
+bool CConnman::ShouldTryToRecv(NodeId node_id) const
+{
+    AssertLockNotHeld(m_nodes_mutex);
+
+    CNode* node{GetNodeById(node_id)};
+    if (node == nullptr) {
+        return false;
+    }
+    return !node->fPauseRecv;
+}
+
 Sock::EventsPerSock CConnman::GenerateWaitSockets(Span<CNode* const> nodes)
 {
+    AssertLockNotHeld(m_nodes_mutex);
+
     Sock::EventsPerSock events_per_sock;
 
     for (const auto& sock : m_listen) {
         events_per_sock.emplace(sock, Sock::Events{Sock::RECV});
     }
 
     for (CNode* pnode : nodes) {
-        bool select_recv = !pnode->fPauseRecv;
-        bool select_send;
-        {
-            LOCK(pnode->cs_vSend);
-            // Sending is possible if either there are bytes to send right now, or if there will be
-            // once a potential message from vSendMsg is handed to the transport. GetBytesToSend
-            // determines both of these in a single call.
-            const auto& [to_send, more, _msg_type] = pnode->m_transport->GetBytesToSend(!pnode->vSendMsg.empty());
-            select_send = !to_send.empty() || more;
-        }
+        const bool select_recv{ShouldTryToRecv(pnode->GetId())};
```

Comment:
dc6393cb93c4851a363b69fd474656cac1ae3b3b: it would be good to annotate in `CConnman::SocketHandlerConnected` above `sendSet = it->second.occurred & Sock::SEND;` that `Sock::SEND` is unset when `ShouldTryToRecv` is `false`.

Reply:
Hmm, you probably mean `ShouldTryToSend` is `false`, right? And then the same applies to receiving. Amended this:

```diff
-                recvSet = it->second.occurred & Sock::RECV;
-                sendSet = it->second.occurred & Sock::SEND;
+                recvSet = it->second.occurred & Sock::RECV; // Sock::RECV could only be set if ShouldTryToRecv() has returned true in GenerateWaitSockets().
+                sendSet = it->second.occurred & Sock::SEND; // Sock::SEND could only be set if ShouldTryToSend() has returned true in GenerateWaitSockets().
```